packages/cli/src/controllers/users.controller.ts

```diff
@@ -157,131 +161,167 @@ export class UsersController {
+					ownedCredentialIds,
+					transfereePersonalProject.id,
+					trx,
+				);
 
-				// Prevents issues with unique key constraints since user being assigned
-				// workflows and credentials might be a sharee
+				// Delete the relationship to the user to delete.
+				// If we don't do this then the following code that deletes the user
+				// will also delete these workflows and credentials.
+				await this.sharedWorkflowRepository.deleteByIds(
+					ownedWorkflowIds,
+					personalProjectToDelete.id,
+					trx,
+				);
 				await this.sharedCredentialsRepository.deleteByIds(
-					transactionManager,
-					sharedCredentialIds,
-					transfereePersonalProject,
+					ownedCredentialIds,
+					personalProjectToDelete.id,
+					trx,
 				);
 
-				// Transfer ownership of owned credentials
-				await transactionManager.update(
-					SharedCredentials,
-					{ project: personalProject, role: 'credential:owner' },
-					{ project: transfereePersonalProject },
+				// Transfer relationships that are not `*:owner`.
+				// This will NOT override whatever relationship the transferee already
+				// has to the resource at the moment.
+
+				const sharedWorkflowIdsOfTransferee = allSharedWorkflows
+					.filter((sw) => sw.projectId === transfereePersonalProject.id)
+					.map((sw) => sw.workflowId);
+				const sharedCredentialIdsOfTransferee = allSharedCredentials
+					.filter((sc) => sc.projectId === transfereePersonalProject.id)
+					.map((sc) => sc.credentialsId);
+
+				// all resources that are shared with the user to delete, but not with the transferee
+				const sharedWorkflowsToTransfer = sharedWorkflowsOfUserToDelete.filter(
+					(sw) =>
+						sw.role !== 'workflow:owner' && !sharedWorkflowIdsOfTransferee.includes(sw.workflowId),
+				);
+				const sharedCredentialsToTransfer = sharedCredentialsOfUserToDelete.filter(
+					(sc) =>
+						sc.role !== 'credential:owner' &&
+						!sharedCredentialIdsOfTransferee.includes(sc.credentialsId),
 				);
 
-				await transactionManager.delete(AuthIdentity, { userId: userToDelete.id });
-
-				// This will remove all shared workflows and credentials not owned
-				await transactionManager.delete(User, { id: userToDelete.id });
-			});
-
-			void this.internalHooks.onUserDeletion({
-				user: req.user,
-				telemetryData,
-				publicApi: false,
+				await Promise.all([
+					trx.insert(
+						SharedWorkflow,
+						sharedWorkflowsToTransfer.map((sw) => ({
+							workflowId: sw.workflowId,
+							projectId: transfereePersonalProject.id,
+							role: sw.role,
+						})),
+					),
+
+					trx.insert(
+						SharedCredentials,
+						sharedCredentialsToTransfer.map((sc) => ({
+							credentialsId: sc.credentialsId,
+							projectId: transfereePersonalProject.id,
+							role: sc.role,
+						})),
+					),
+				]);
 			});
-			await this.externalHooks.run('user.deleted', [await this.userService.toPublic(userToDelete)]);
-			return { success: true };
```

Comment:
That's part of the refactor:

> the deletion logic is is not duplicated anymore (once for the transfer case and once for not transfering), instead it will first transfer everything **and then delete the user like it would anyways**,

Migrating the data does not need to duplicate the business logic for deletion. The concerns are completely separated now within the function. If the need arises we can also separate the the resource migration logic out. I don't think it's necessary to do this right now, but I've heard that it may come in the future (migrating workflows/credentials from one user to another).